packages/tasit-sdk/src/Decentraland.test.js

```diff
@@ -249,393 +314,657 @@ describe("Decentraland", () => {
             [SMALL_AMOUNT]
           );
 
-          const toAddress = ephemeralAddress;
+          // Set wallet and signers
+          gnosisSafe.setSigners([gnosisSafeOwner]);
+          gnosisSafe.setWallet(gnosisSafeOwner);
+        });
 
-          // Transfer a small amount of ethers to ephemeral account to pay for gas
-          const transferEthersAction = gnosisSafe.transferEther(
-            toAddress,
-            SMALL_AMOUNT
-          );
-          await transferEthersAction.send();
-          await transferEthersAction.waitForOneConfirmation();
-          await expectExactEtherBalances(provider, [toAddress], [SMALL_AMOUNT]);
-
-          const transferManaAction = gnosisSafe.transferERC20(
-            MANA_ADDRESS,
-            toAddress,
-            manaAmountForShopping
-          );
-          await transferManaAction.send();
-          await transferManaAction.waitForOneConfirmation();
-          await expectExactTokenBalances(
-            mana,
-            [toAddress],
-            [manaAmountForShopping]
-          );
+        // Transfer a small amount of ethers to ephemeral account to pay for gas
+        beforeEach("onboarding - funding ephemeral wallet with ETH", done => {
+          (async () => {
+            const toAddress = ephemeralAddress;
 
-          mana.setWallet(ephemeralWallet);
-          const approvalAction = mana.approve(
-            MARKETPLACE_ADDRESS,
-            manaAmountForShopping
-          );
-          await approvalAction.send();
-          await approvalAction.waitForOneConfirmation();
+            const action = gnosisSafe.transferEther(toAddress, SMALL_AMOUNT);
 
-          const allowance = await mana.allowance(
-            ephemeralAddress,
-            MARKETPLACE_ADDRESS
-          );
+            const confirmationListener = async () => {
+              await expectExactEtherBalances(
+                provider,
+                [toAddress],
+                [SMALL_AMOUNT]
+              );
 
-          expect(`${allowance}`).to.equal(`${manaAmountForShopping}`);
-        });
+              done();
+            };
+
+            const errorListener = error => {
+              action.off("error");
+              done(error);
+            };
+
+            action.once("confirmation", confirmationListener);
+            action.on("error", errorListener);
 
-        it("should buy an estate", async () => {
-          const {
-            assetId,
-            nftAddress,
-            seller,
-            priceInWei,
-            expiresAt,
-          } = estateForSale;
+            action.send();
+          })();
+        });
 
-          await checkAsset(estate, mana, estateForSale, ephemeralAddress);
+        beforeEach("onboarding - funding ephemeral wallet with MANA", done => {
```

Comment:
Out of curiosity, why not have multiple things happen in a single `beforeEach`?

Reply:
Yes, it's related to https://github.com/tasitlabs/tasit/issues/151#issuecomment-474908828 discussion.
I've tried to avoid "callback hell" since these actions are nested.

Reply:
Ah ok. Do these `beforeEach`'s happen sequentially or in parallel.

The other way to avoid "callback hell" in a single `beforeEach` would be to ship marketplace approval and mana funding in parallel, have a separate listener for each that sets local state for whether it has it's response yet, and each time when setting state check if the other one is already done, and if so finish the whole before each.

Is the only callback hell here that we wouldn't want to ship that double action only once the single action of receiving eth had completed?

```js
// buildManaAction here
// buildEthAction here
// buildMarketplaceAction here

functionThatSetsManaDone {
  if (marketplaceDone) {
    functionThatHandlesMarketplaceAndManaDone()
  }
}

functionThatShipsManaAndMarketplaceApproval {
  manaAction.send()
  marketplaceAction.send()
}

ethAction.on('confirmation', functionThatShipsManaAndMarketplaceApproval)
manaAction.on('confirmation', functionThatSetsManaDone)
marketplaceAction.on('confirmation', functionThatSetsMarketplaceDone)

ethAction.send()
```

I guess it is a little confusing to reason about, but as long as we extract the function and define it ahead of time, we're not actually defining functions in a nested way that creates lots of indentation, although there is nesting of program flow.

And we could of course listen for events rather than confirmations above

Reply:
Extracted to: https://github.com/tasitlabs/TasitSDK/issues/381

Reply:
Thanks. Resolving this.